pkg/operator/controller/ingress/status.go

```diff
@@ -171,35 +181,109 @@ func computeDeploymentPodsScheduledCondition(deployment *appsv1.Deployment, pods
+		},
+	}
+	_, unavailableConditions := checkAnnotatedConditions(expected, conditionsMap)
+
+	// TODO - check with team - do we want to use a grace period when checking Available?
+
+	if len(unavailableConditions) != 0 {
+		degraded := formatConditions(unavailableConditions)
+		return []operatorv1.OperatorCondition{
+			{
+				Type:    operatorv1.IngressControllerAvailableConditionType,
+				Status:  operatorv1.ConditionFalse,
+				Reason:  "IngressControllerUnavailable",
+				Message: "One or more status conditions indicate unavailable: " + degraded,
+			},
 		}
-		switch cond.Status {
-		case corev1.ConditionTrue:
-			return operatorv1.OperatorCondition{
+	} else {
+		return []operatorv1.OperatorCondition{
+			{
 				Type:   operatorv1.IngressControllerAvailableConditionType,
 				Status: operatorv1.ConditionTrue,
-			}
-		case corev1.ConditionFalse:
-			return operatorv1.OperatorCondition{
-				Type:    operatorv1.IngressControllerAvailableConditionType,
-				Status:  operatorv1.ConditionFalse,
-				Reason:  cond.Reason,
-				Message: "The deployment is unavailable: " + cond.Message,
-			}
+			},
 		}
 	}
+}
 
-	return operatorv1.OperatorCondition{
-		Type:    operatorv1.IngressControllerAvailableConditionType,
-		Status:  operatorv1.ConditionFalse,
-		Reason:  "DeploymentAvailabilityUnknown",
-		Message: "The deployment's Available condition couldn't be interpreted",
+func checkAnnotatedConditions(expecteds []expectedCondition, conditionsMap map[string]*operatorv1.OperatorCondition) ([]*operatorv1.OperatorCondition, []*operatorv1.OperatorCondition) {
+	var graceConditions, degradedConditions []*operatorv1.OperatorCondition
+	now := clock.Now()
+	// Keep checking conditions every minute while degraded.
+	requeueAfter := time.Minute
+	for _, expected := range expecteds {
+		condition, haveCondition := conditionsMap[expected.condition]
+		if !haveCondition {
+			if expected.mustHave {
+				degradedConditions = append(degradedConditions, &operatorv1.OperatorCondition{Type: expected.condition, Status: operatorv1.ConditionUnknown})
+			}
+			continue
+		}
+		if condition.Status == expected.status {
+			continue
+		}
+		failedPredicates := false
+		for _, ifCond := range expected.ifConditionsTrue {
+			predicate, havePredicate := conditionsMap[ifCond]
+			if !havePredicate || predicate.Status != operatorv1.ConditionTrue {
+				failedPredicates = true
+				break
+			}
+		}
+		if failedPredicates {
+			continue
+		}
+		if expected.gracePeriod != 0 {
+			t1 := now.Add(-expected.gracePeriod)
+			t2 := condition.LastTransitionTime
+			if t2.After(t1) {
+				d := t2.Sub(t1)
+				if len(graceConditions) == 0 || d < requeueAfter {
+					// Recompute status conditions again
+					// after the grace period has elapsed.
+					requeueAfter = d
+				}
+				graceConditions = append(graceConditions, condition)
+				continue
+			}
+		}
+		degradedConditions = append(degradedConditions, condition)
 	}
+	return graceConditions, degradedConditions
```

Comment:
Need to return `requeueAfter` for `computeIngressDegradedCondition`.

Reply:
Ouch, that would have been bad.